packages/mina-service/src/domain/rollup-offchain.ts

```diff
@@ -0,0 +1,135 @@
+import { deserializeTransition } from '@helper';
+import { EProofStatusDocument } from '@zkdb/common';
+import { ZkDbProcessor } from '@zkdb/smart-contract';
+import {
+  getCurrentTime,
+  ModelMetadataDatabase,
+  ModelRollupOffChain,
+  ModelTransitionLog,
+  TCompoundSession,
+  TRollupQueueData,
+} from '@zkdb/storage';
+import { MerkleTree } from 'o1js';
+
+const MAX_MERKLE_TREE_HEIGHT = 256;
+const MIN_MERKLE_TREE_HEIGHT = 8;
+
+export class RollupOffchain {
+  static async create(task: TRollupQueueData, session: TCompoundSession) {
+    const { serverless, proofService } = session;
+    const { status, databaseName } = task;
+    // Ensure status must be 'queued' to process
+    if (status !== EProofStatusDocument.Queued) {
+      throw new Error(
+        `Task status should be in ${EProofStatusDocument.Queued}`
+      );
+    }
+
+    const imMetadataDatabase = ModelMetadataDatabase.getInstance();
+
+    const metadataDatabase = await imMetadataDatabase.findOne(
+      { databaseName },
+      { session: serverless }
+    );
+
+    if (!metadataDatabase) {
+      throw new Error(`Metadata database ${databaseName} cannot be found`);
+    }
+
+    const { merkleHeight } = metadataDatabase;
+
+    if (
+      merkleHeight < MIN_MERKLE_TREE_HEIGHT ||
+      merkleHeight > MAX_MERKLE_TREE_HEIGHT
+    ) {
+      throw new Error(
+        `Invalid Merkle height. Expected 8-256, got ${merkleHeight}`
+      );
+    }
+
+    // ZkDbProcessor will automatically compile when getInstance
+    const zkAppProcessor = await ZkDbProcessor.getInstance(merkleHeight);
+
+    const imRollupOffChain = ModelRollupOffChain.getInstance();
+
+    // Get previous proof to update
+    const previousProof = await imRollupOffChain.findOne(
```

Comment:
We need a constraint between previous proof and transition, if you get wrong transition the proof is also wrong.